File: code-examples/transaction-monitoring/backend-python/services/compilot.py
import os
import requests
from dotenv import load_dotenv

"""
Service for interacting with the ComPilot API.
Handles transaction submissions and API communication.

Required environment variables:
- COMPILOT_API_URL: Base URL for the ComPilot API
- COMPILOT_API_KEY: API key for authentication
- COMPILOT_TMS_WORKFLOW_ID: Workflow ID for transaction monitoring
"""

load_dotenv()

class ComPilotService:
    @staticmethod
    def submit_transaction(transaction):
        """
        Submits a transaction to the ComPilot API for processing.
        
        Args:
            transaction: Dictionary containing transaction details
            
        Returns:
            Dictionary containing the API response
            
        Raises:
            Exception: If API credentials are missing or if the API request fails
        """
        try:
            workflow_id = os.getenv('COMPILOT_TMS_WORKFLOW_ID')
            if not workflow_id:
                raise Exception('COMPILOT_TMS_WORKFLOW_ID is not defined')

            response = requests.post(
                f"{os.getenv('COMPILOT_API_URL')}/workflows/{workflow_id}/transactions",
                json=transaction,
                headers={
                    'Content-Type': 'application/json',
                    'Authorization': f"Bearer {os.getenv('COMPILOT_API_KEY')}"
                }
            )
            
            response.raise_for_status()
            return response.json()
            
        except requests.exceptions.RequestException as e:
            if e.response is not None:
                error_data = e.response.json()
                raise Exception(f"ComPilot API Error: {error_data}")
            raise Exception(f"ComPilot API Error: {str(e)}") 

File: code-examples/transaction-monitoring/backend-typescript/src/services/compilot.ts
import { Transaction } from '../types/transaction';

/**
 * Service for interacting with the ComPilot API.
 * Handles transaction submissions and API communication.
 * 
 * Required environment variables:
 * - COMPILOT_API_URL: Base URL for the ComPilot API
 * - COMPILOT_API_KEY: API key for authentication
 * - COMPILOT_TMS_WORKFLOW_ID: Workflow ID for transaction monitoring
 */
export class ComPilotService {
    /**
     * Submits a transaction to the ComPilot API for processing.
     * 
     * @param transaction - The transaction details to submit
     * @throws Error if API credentials are missing or if the API request fails
     * @returns The API response data
     * 
     * Example usage:
     * ```typescript
     * const transaction = {
     *   transactionType: 'crypto',
     *   transactionSubType: 'wallet transfer',
     *   transactionInfo: {
     *     direction: 'IN',
     *     amount: 0.5,
     *     currencyCode: 'ETH'
     *   }
     * };
     * const response = await ComPilotService.submitTransaction(transaction);
     * ```
     */
    static async submitTransaction(transaction: Transaction) {
        console.log('📝 Environment variables:', {
            apiUrl: process.env.COMPILOT_API_URL,
            hasApiKey: !!process.env.COMPILOT_API_KEY,
            workflowId: process.env.COMPILOT_TMS_WORKFLOW_ID
        });

        if (!process.env.COMPILOT_API_URL) {
            throw new Error('COMPILOT_API_URL is not defined');
        }

        if (!process.env.COMPILOT_TMS_WORKFLOW_ID) {
            throw new Error('COMPILOT_TMS_WORKFLOW_ID is not defined');
        }

        const response = await fetch(
            `${process.env.COMPILOT_API_URL}/workflows/${process.env.COMPILOT_TMS_WORKFLOW_ID}/transactions`, 
            {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${process.env.COMPILOT_API_KEY}`
                },
                body: JSON.stringify(transaction)
            }
        );

        const data = await response.json();

        if (!response.ok) {
            throw new Error(`ComPilot API Error: ${JSON.stringify(data, null, 2)}`);
        }

        return data;
    }
} 